A34-weighted-round-robin.md

```diff
@@ -1,103 +1,101 @@
-*/
-PickResult Pick(PickArgs args)
-```
+This part extends the behavior of EDS section in [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md). Instead of discarding the per endpoint `load_balancing_weight`, we want to add it to per-address attibute and pass it along to `lb_policy`.
 
-### On new `ClusterLoadAssignment`
+#### `lb_policy` for per endpoint `load_balancing_weight` from `ClusterLoadAssignment`
+When the `lb_policy` field in CDS response is `ROUND_ROBIN`, we use `weighted_round_robin` as the `lb_policy`.
 
-#### endpoints have weight change
-- Update EDF priority queue by updating entries have weight change.
-- Different from [Envoy](https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/upstream/load_balancer_impl.cc#L733-L737), we'd like to udpate EDF priority queue so that new weights applied immediately even endpoints list is not changed.
+As of today, we only accept `ROUND_ROBIN` as `lb_policy` in CDS response per [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md). Therefore, `weighted_round_robin` will always be used.
 
-#### endpoints list change
-- Udpate EDF priority queue by adding/removing entries to reflect the change immediately.
+#### On update of `ClusterLoadAssignment`
+When an EDS update is received, an update will be sent to the `lb_policy`. The `lb_policy` will create a new picker. This is slightly Different from [Envoy](https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/upstream/load_balancer_impl.cc#L733-L737). We'd like to udpate EDF priority queue so that new weights applied immediately even endpoints list is not changed.
 
-### On endpoint ConnectiveState Update
-- Udpate EDF priority queue by adding/removing entries to reflect the change immediately.
+#### NOTE
+- `weighted_round_robin` should always be updated to the lastest `ClusterLoadAssignment`. It's xDS server's responsibility to maintain consistency.
 
 ## Rationale
 
 Several applications can be built upon this feature, e.g. utilization load balancing, blackhole erroring endpoints, load testing,... etc.
 
 The reason to refresh EDF picker even there is only weight change on some endpoints which is different from envoy is because we'd like real time traffic shift for use cases like load testing, blackhole erroring endpoints.
 
-The reasons to introduce a new algorithm instead of using existing `weighted_target` policy are
+The reasons to introduce a new algorithm instead of re-using the same algorithm of `weighted_target` policy are
 - [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) maintains FIFO order for endpoints with same weight which is easier to reason.
 - We want to be consistent with the behavior of Envoy.
```

Comment:
There is one important architectural difference between Envoy and gRPC to be aware of here.  In Envoy, the LB policy doesn't handle connection management; it just chooses from the set of configured hosts, and once it does that, Envoy figures out if it has an existing connection to that host or if it needs to create a new one.  But in gRPC, the LB policy is responsible for deciding when to connect to the endpoints.  (The gRPC LB policy also needs to monitor the connectivity state of all subchannels, and it must tell the channel to use only subchannels that are actually connected.  This generally means that the LB policy needs to create a new picker whenever the connectivity state of a subchannel changes.)

The `round_robin` LB policy's connection management behavior is that it always tries to keep one connection open to each address in the address list at all times.  When the policy is first instantiated, it immediately tries to connect to all addresses, and whenever a subchannel becomes disconnected, it immediately tries to reconnect.

This behavior is different from how Envoy's connection pools work.  But given that the main purpose of the `round_robin` policy is to spread load evenly across the backends, that implies that the client will be generating a fair amount of traffic, which implies that proactively connecting this way is generally better.  The Envoy approach actually causes increased latency, because the first RPC sent to each endpoint winds up needing to wait for it to connect, even if there's an existing connection to a different endpoint that could handle the traffic.

I think that if we want this new policy to be a drop-in replacement for the existing `round_robin` policy, we'll need to use this same connection management behavior.

Reply:
Thanks, this is really helpful. I added a section `Subchannel connectivity management`. I just slightly modify how rr manages connectivity from this comment. Hope there is no copyright issue : )